cli/js/errors.ts

```diff
@@ -181,24 +181,55 @@ class Busy extends Error {
     this.name = "Busy";
   }
 }
+/* eslint-disable @typescript-eslint/no-explicit-any */
+class AssertionError extends Error {
+  #actual: any;
+  #expected: any;
+  constructor(
+    message?: string,
+    { actual, expected }: { actual?: any; expected?: any } = {}
+  ) {
+    super(message);
+    this.name = "AssertionError";
+    this.#actual = actual;
+    this.#expected = expected;
+  }
+
+  get actual(): any {
+    return this.#actual;
+  }
+
+  get expected(): any {
+    return this.#expected;
+  }
+}
+/* eslint-enable */
+class NotImplemented extends Error {
+  constructor(msg: string) {
+    super(msg);
+    this.name = "NotImplemented";
+  }
+}
 
 export const errors = {
-  NotFound: NotFound,
-  PermissionDenied: PermissionDenied,
-  ConnectionRefused: ConnectionRefused,
-  ConnectionReset: ConnectionReset,
-  ConnectionAborted: ConnectionAborted,
-  NotConnected: NotConnected,
-  AddrInUse: AddrInUse,
-  AddrNotAvailable: AddrNotAvailable,
-  BrokenPipe: BrokenPipe,
-  AlreadyExists: AlreadyExists,
-  InvalidData: InvalidData,
-  TimedOut: TimedOut,
-  Interrupted: Interrupted,
-  WriteZero: WriteZero,
-  UnexpectedEof: UnexpectedEof,
-  BadResource: BadResource,
-  Http: Http,
-  Busy: Busy,
+  NotFound,
+  PermissionDenied,
+  ConnectionRefused,
+  ConnectionReset,
+  ConnectionAborted,
+  NotConnected,
+  AddrInUse,
+  AddrNotAvailable,
+  BrokenPipe,
+  AlreadyExists,
+  InvalidData,
+  TimedOut,
+  Interrupted,
+  WriteZero,
+  UnexpectedEof,
+  BadResource,
+  Http,
+  Busy,
+  AssertionError,
+  NotImplemented,
```

Comment:
I don't know about exposing AssertionError and NotImplemented as basic error codes. This isn't a list that should be expanded much - it corresponds to "errno" in posix.

NotImplemented in particular, I'm not even sure should be a recoverable error. It would probably be better to crash the process.

AssertionError seems odd because assertions are not part of the API. Is it a web API? Should it be a global?

Reply:
> I don't know about exposing AssertionError and NotImplemented as basic error codes. This isn't a list that should be expanded much - it corresponds to "errno" in posix.

We could put them on `Deno` but it seemed less confusing to have them with the other thing called `errors`.

> NotImplemented in particular, I'm not even sure should be a recoverable error. It would probably be better to crash the process.

Really? What about the symlink errors on Windows?  Some of the unimplemented fetch API?  Seems like if you are writing isomorphic code, you would want/need to recover from such errors.

> AssertionError seems odd because assertions are not part of the API. Is it a web API? Should it be a global?

Maybe they should be (ref: #4090).  Having it as just a plain global isn't part of any spec that I am aware of, though as you know Node.js has it as a global and `assert.AssertionError` is the location of error in Node.js.  That pattern has been adopted by others.

From a non-public place, a lot of the WHATWG specs (like streams: https://streams.spec.whatwg.org/) and I suspect it is in other spec stuff, that you have to make assertions, just like we do in a lot of our internal our internal code.  If such an assertion fails, we should throw something that isn't `Error`.  If we throw something other than `Error` it seems cruel not to expose that.